### PR TITLE
Schema for macro definitions

### PR DESCRIPTION
**Pull request by ReaLNeroM**
In this PR, get_schema_for_macro_definitions is implemented. 
Here's my methodology: 

- We check for non-default directives, and report an error.
- We add the directives required for a macro edge definition(@macro_edge_definition and @macro_edge-target). 
- Then any directives not in DIRECTIVES_ALLOWED_IN_MACRO_EDGE_DEFINITION are removed from the set.
- We create a shallow copy of the schema, where the directives list is replaced by a list containing the elements of the set. @reviewers: Could the shallow copy cause issues, if for example, the macro editor modifies the schema? Is this a vulnerability? EDIT: Updated the function docstring to reflect modifying the schema is **undefined behavior**

### CODE REVIEW

File: graphql_compiler/macros/__init__.py [hunk 1]
```diff
@@ -167,16 +170,26 @@ def get_schema_for_macro_definition(schema):
...
+    like @output and @output_source, will be removed from the directives list.
 
     Args:
         schema: GraphQLSchema over which we want to write macros
 
     Returns:
         GraphQLSchema usable for writing macros
     """
-    raise NotImplementedError()  # TODO(bojanserafimov): Implement
+    macro_definition_schema = copy(schema)
```
Comment (pmantica1) [inline graphql_compiler/macros/__init__.py]: The copying seems fine to me. I don't understand what you mean by vulnerabilities.

File: graphql_compiler/macros/__init__.py [hunk 2]
```diff
@@ -167,16 +170,26 @@ def get_schema_for_macro_definition(schema):
...
+    macro_definition_schema_directives_set = set(macro_definition_schema.get_directives()).union(
+        DIRECTIVES_REQUIRED_IN_MACRO_EDGE_DEFINITION)
+    # Exclude disallowed directives
+    macro_definition_schema_directives_set &= DIRECTIVES_ALLOWED_IN_MACRO_EDGE_DEFINITION
```
Comment (pmantica1) [inline graphql_compiler/macros/__init__.py]: I think it's better to validate that all of the current schema directives are allowed and raise a validation error if one of them isn't. (Look for the exception that fits best). Then I would add a test that validation works and update the docstring with a "Raises" block.

Comment (pmantica1) [reply graphql_compiler/macros/__init__.py]: If you would continue with your current approach I would separate this into functions: one to remove the unsupported directives and another to add the required directives.

Comment (bojanserafimov) [reply graphql_compiler/macros/__init__.py]: Agree with Pedro on raising an exception. The macro system is not a macro system for any graphql schema, but specifically a macro system for this graphql compiler. If it's being used with anything else, it should not promise to work and pretend that everything is ok.

File: graphql_compiler/macros/__init__.py (hunk 1 above)
Comment (ReaLNeroM) [reply graphql_compiler/macros/__init__.py]: The returned schema's fields are the same copy as the argument schema's fields. So if a macro editor edits the schema returned from this function, we might run into issues. How can we make the fields immutable?

File: graphql_compiler/macros/__init__.py (hunk 2 above)
Comment (ReaLNeroM) [reply graphql_compiler/macros/__init__.py]: Added schema.py/_check_for_nondefault_macro_directives() to detect this issue.

File: graphql_compiler/macros/__init__.py [hunk 3]
```diff
@@ -167,16 +172,30 @@ def get_schema_for_macro_definition(schema):
...
         GraphQLSchema usable for writing macros
     """
-    raise NotImplementedError()  # TODO(bojanserafimov): Implement
+
+    macro_definition_schema = copy(schema)
+    macro_definition_schema_directives = schema.get_directives()
+    _check_for_nondefault_macro_directives(macro_definition_schema_directives)
+    macro_definition_schema_directives = include_required_macro_directives(
+        macro_definition_schema_directives)
+    macro_definition_schema_directives = exclude_disallowed_directives_in_macros(
```
Comment (pmantica1) [inline graphql_compiler/macros/__init__.py]: Why are we excluding disallowed directives? Didn't Bojan and I agree on just asserting that we simply raise an error if any of schema directives is not allowed?

Comment (ReaLNeroM) [reply graphql_compiler/macros/__init__.py]: Some default directives, like @output and @output_source are disallowed. However, if we raised errors, we wouldn't be able to use @output and macro edges in the same schema.

File: graphql_compiler/schema.py [hunk 4]
```diff
@@ -349,3 +349,15 @@ def insert_meta_fields_into_existing_schema(graphql_schema):
                                      .format(meta_field_name))
 
             type_obj.fields[meta_field_name] = meta_field
+
+
+def _check_for_nondefault_macro_directives(directives):
```
[the remark this answers is not in the document]
Comment (ReaLNeroM) [reply graphql_compiler/schema.py]: The function can either leave a directive as-is, remove it from the schema(@output), or raise an error(any user-created directive). Since we don't want a DEFAULT_DIRECTIVES_DISALLOWED_IN_MACRO_EDGE_DEFINITION per [this comment](https://github.com/kensho-technologies/graphql-compiler/pull/313#commitcomment-33655951), we need whether all directives are default.

File: graphql_compiler/schema.py [hunk 5]
```diff
@@ -349,3 +349,17 @@ def insert_meta_fields_into_existing_schema(graphql_schema):
                                      .format(meta_field_name))
 
             type_obj.fields[meta_field_name] = meta_field
+
+
+def _check_for_nondefault_directives(directives):
+    """Check if any user-created directives are present"""
+    deprecated_directive_names = {'deprecated', 'skip', 'include'}
```
Comment (bojanserafimov) [inline graphql_compiler/schema.py]: These directives are not deprecated. They're the default.

File: graphql_compiler/schema.py [hunk 6]
```diff
@@ -349,3 +349,17 @@ def insert_meta_fields_into_existing_schema(graphql_schema):
...
+    default_directive_names = {directive.name for directive in DIRECTIVES}
```
Comment (bojanserafimov) [inline graphql_compiler/schema.py]: Might be a good idea to check these for equality, instead of just comparing names. It's possible that someone defines a different `@optional` directive for example. Not likely to happen, but if it does it will be difficult to find because we think we are addressing the problem.

File: graphql_compiler/tests/test_macro_schema.py [hunk 7]
```diff
@@ -32,3 +37,115 @@ def test_get_schema_with_macros_basic(self):
...
+            }''',
+            '''{
+                Animal @macro_edge_definition(name: "out_Animal_RelatedEntity") {
+                    in_Entity_Related {
+                        ... on Entity @macro_edge_target {
+                            uuid
+                        }
+                    }
+                }
+            }''']
```
Comment (bojanserafimov) [inline graphql_compiler/tests/test_macro_schema.py]: You can import these from `test_helpers.py`

File: graphql_compiler/tests/test_macro_schema.py [hunk 8]
```diff
@@ -32,3 +37,115 @@ def test_get_schema_with_macros_basic(self):
...
+        for directive in original_schema.get_directives():
+            if directive in DIRECTIVES_ALLOWED_IN_MACRO_EDGE_DEFINITION:
+                self.assertTrue(directive in macro_definition_schema.get_directives())
+
+    def test_get_schema_for_macro_definition_removal(self):
+        schema_with_macros = get_schema_with_macros(self.macro_registry)
+        macro_definition_schema = get_schema_for_macro_definition(schema_with_macros)
+        for directive in macro_definition_schema.get_directives():
+            self.assertTrue(directive.name != '@output')
+            self.assertTrue(directive.name != '@output_source')
```
Comment (bojanserafimov) [inline graphql_compiler/tests/test_macro_schema.py]: `directive.name` is a `Name` object. You can get it's string contents by calling `directive.name.value`. If you cargo-culted this code from somewhere else, please check that we are not making the same mistake there too :) It's possible.

Comment (bojanserafimov) [inline graphql_compiler/tests/test_macro_schema.py]: The `@` symbol will not be in the directive name. You can use `OutputSourceDirective.name` to avoid magic strings, and having to remember this.

File: graphql_compiler/schema.py (hunk 6 above)
Comment (ReaLNeroM) [reply graphql_compiler/schema.py]: Directive definitions include a description, but the argument directives do not. If we check for equality, the descriptions get compared as well. 
What if we just create a copy of DIRECTIVES, with descriptions nulled out? What about deprecated, skip, and include then?
It's way too adversarial to assume users will create their own directives, with the same name as the ones we've reserved.

File: graphql_compiler/tests/test_macro_schema.py (hunk 8 above)
Comment (ReaLNeroM) [reply graphql_compiler/tests/test_macro_schema.py]: directive.name is a str. Check again.

File: graphql_compiler/schema.py (hunk 6 above)
Comment (ReaLNeroM) [reply graphql_compiler/schema.py]: Changed the function name to reflect it only compares names.
Get_schema_for_macro_definitons now has a precondition assuming all directives we support like @fold, @output, etc. have their default behavior.

File: graphql_compiler/schema.py [hunk 9]
```diff
@@ -291,7 +292,6 @@ def _parse_datetime_value(value):
     FoldDirective,
 )
 
-
```
Comment (pmantica1) [inline graphql_compiler/schema.py]: nit: Please don't change line formatting that is unrelated to this pr.

File: graphql_compiler/tests/test_macro_schema.py (hunk 8 above)
Comment (bojanserafimov) [reply graphql_compiler/tests/test_macro_schema.py]: Ah, correct. It's only a `Name` in directive instances.
